pytket/extensions/cutensornet/structured_state/mps.py

```diff
@@ -160,48 +160,86 @@ def apply_gate(self, gate: Command) -> MPS:
+                unitaries are supported.
+
+        Returns:
+            ``self``, to allow for method chaining.
+
+        Raises:
+            RuntimeError: If the ``CuTensorNetHandle`` is out of scope.
+            ValueError: If the number of qubits provided is not one or two.
+            ValueError: If the size of the matrix does not match with the number of
+                qubits provided.
+            ValueError: If qubits are non-adjacent.
         """
         if self._lib._is_destroyed:
             raise RuntimeError(
                 "The cuTensorNet library handle is out of scope.",
                 "See the documentation of update_libhandle and CuTensorNetHandle.",
             )
 
-        positions = [self.qubit_position[q] for q in gate.qubits]
-        if any(self.get_physical_dimension(pos) != 2 for pos in positions):
-            raise RuntimeError(
-                "Gates can only be applied to tensors with physical"
-                + " bond dimension of 2."
-            )
-        self._logger.debug(f"Applying gate {gate}")
+        self._logger.debug(f"Applying unitary {unitary} on {qubits}.")
+
+        if len(qubits) == 1:
+            if unitary.shape != (2, 2):
+                raise ValueError(
+                    "The unitary introduced acts on one qubit but it is not 2x2."
+                )
 
-        if len(positions) == 1:
-            self._apply_1q_gate(positions[0], gate.op)
+            self._apply_1q_unitary(unitary, qubits[0])
             # NOTE: if the tensor was in canonical form, it remains being so,
             #   since it is guaranteed that the gate is unitary.
 
-        elif len(positions) == 2:
+        elif len(qubits) == 2:
+            if unitary.shape != (4, 4):
+                raise ValueError(
+                    "The unitary introduced acts on two qubits but it is not 4x4."
+                )
+
+            positions = [self.qubit_position[q] for q in qubits]
             dist = positions[1] - positions[0]
             # We explicitly allow both dist==1 or dist==-1 so that non-symmetric
             # gates such as CX can use the same Op for the two ways it can be in.
             if dist not in [1, -1]:
-                raise RuntimeError(
-                    "Gates must be applied to adjacent qubits! "
-                    + f"This is not satisfied by {gate}."
-                )
-            self._apply_2q_gate((positions[0], positions[1]), gate.op)
+                raise ValueError("Gates must be applied to adjacent qubits!")
+            self._apply_2q_unitary(unitary, qubits[0], qubits[1])
             # The tensors will in general no longer be in canonical form.
             self.canonical_form[positions[0]] = None
             self.canonical_form[positions[1]] = None
 
         else:
-            raise RuntimeError(
-                "Gates must act on only 1 or 2 qubits! "
-                + f"This is not satisfied by {gate}."

```

Comment:
Is it intended that you remove the additional information of the gate?

Reply:
Hmm, I could just make that check in `apply_gate` as well, and raise the error with the info about `gate`. I'll do that.

If the error is raised within `get_unitary` I don't think it's that helpful for the error message to print the matrix. For debugging purposes `_logger.debug` provides a good amount of info.

Reply:
Done!